core/src/main/java/org/apache/brooklyn/core/effector/composite/AbstractCompositeEffector.java

```
@@ -0,0 +1,184 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.Beta;
+import com.google.common.collect.Iterables;
+
+@Beta
+public abstract class AbstractCompositeEffector extends AddEffector {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractCompositeEffector.class);
+
+    public AbstractCompositeEffector(Effector<?> effector) {
+        super(effector);
+    }
+
+    @Override
+    public void apply(EntityLocal entity) {
+        Maybe<Effector<?>> effectorMaybe = entity.getEntityType().getEffectorByName(effector.getName());
+        if (!effectorMaybe.isAbsentOrNull()) {
+//            Effector<?> original = Effectors.effector(effectorMaybe.get()).name(ORIGINAL_PREFIX + effector.getName()).build();
+//            ((EntityInternal) entity).getMutableEntityType().addEffector(original);
+        }
+        super.apply(entity);
+    }
+
+    protected static abstract class Body<T> extends EffectorBody<T> {
+        protected final Effector<?> effector;
+        protected final ConfigBag params;
+
+        public Body(Effector<?> eff, ConfigBag params) {
+            this.effector = eff;
+            this.params = params;
+        }
+
+        @Override
+        public abstract T call(final ConfigBag params);
+
+        protected String getEffectorName(Object effectorDetails) {
+            String effectorName = null;
+            if (effectorDetails instanceof String) {
+                effectorName = (String) effectorDetails;
+            } else if (effectorDetails instanceof Map) {
+                Map<String,Object> effectorMap = (Map) effectorDetails;
+                Set<String> keys = effectorMap.keySet();
+                if (keys.size() != 1) {
+                    throw new IllegalArgumentException("Effector parameter cannot be parsed: " + effectorDetails);
+                }
+                effectorName = Iterables.getOnlyElement(keys);
+            } else {
```

Indeed, it would clearly be a mistake - prefer to fail fast, so perhaps throwing an exception rather than allowing it to pass by calling `toString()`